### PR TITLE
Always use canonical urls in the sitemap

### LINKED ISSUE #23363: New issues with product url's in sitemap.xml generation after PR 23129 got merged.
### Preconditions (*)
1. Magento `2.3-develop` branch after https://github.com/magento/magento2/pull/23129 got merged, I've used commit 51773443eae to test this again

### Steps to reproduce (*)
1. Create a category with url key 'category'
2. Create a product with url key 'product' and assign it to the category you created in step 1
3. Go to Stores > Configuration > Catalog > Catalog > Search Engine Optimization

#### First problem:
1. Make sure the following settings are setup:
    - Use Categories Path for Product URLs: Yes
    - Generate "category/product" URL Rewrites: Yes
    - Use Canonical Link Meta Tag For Products: Yes
2. Flush caches
3. In Marketing > SEO & Search > Site Map, create a sitemap and look at the generated file.

#### Second problem:
1. Make sure the following settings are setup:
    - Use Categories Path for Product URLs: Yes
    - Generate "category/product" URL Rewrites: No
    - Use Canonical Link Meta Tag For Products: No
2. Flush caches
3. In Marketing > SEO & Search > Site Map, create a sitemap and look at the generated file.

#### First problem:
1. A product link: `https://{domain}/product.html`

#### Second problem:
1. A product link: `https://{domain}/category/product.html`

#### First problem:
1. A product link: `https://{domain}/category/product.html`

#### Second problem:
1. A product link: `https://{domain}/product.html`

### Discussion
This is a follow up ticket from my comments in https://github.com/magento/magento2/pull/23129 after it got (incorrectly?) approved and merged

Please first verify if my claims are correct before starting to work on this. I'm not an SEO expert, I just base this on how I _assume_ search engines work when indexing a shop.

#### First problem
As far as I know, the `sitemap.xml` files' only purpose is for search engines to use as they can then more easily find all important links to index a webshop.
In this case, where `Use Canonical Link Meta Tag For Products` is set to `Yes`, we already indicate on product detail pages that search engines have to use the canonical url (which doesn't contain the category path) using a `<meta>` tag.
But after https://github.com/magento/magento2/pull/23129 got merged, the sitemap files contain product url's with the category path and not the canonical url's. So search engines will first find the non-canonical url, then look at the source code of that page, then find the canonical url, and then visit that url and remove the first page they found in the sitemap file from their index. This means at least 2 requests are made for every product on your shop from search engines which might cause more traffic/load then needed.

#### Second problem
Magento added a new option `Generate "category/product" URL Rewrites` (via [MC-4244](https://github.com/magento/magento2/commit/211dd25c73bb2539c3c064374d9d5649f253fd14)) which is enabled by default, but when you disable it, Magento will no longer store category paths for product url's in the `url_rewrite` table in the database, but still allows you to see url's with the category path on the frontend as those url's are then build dynamically.
Here it is expected that the sitemap files include the links being seen on the frontend even if they don't exist in the `url_rewrite` table, but the sitemap now uses the canonical links instead of the full url including the category path.
This was what https://github.com/magento/magento2/pull/23129 was suppose to fix, but this new option was not taken into consideration.

### PR DESCRIPTION
**Pull request by AntonEvers**
### Description (*)

In Magento 2.3.3, a commit (https://github.com/magento/magento2/commit/e7c2e1f89a91f358548ac2527599684cb1ec8ea8) was introduced by @Vaha that make the sitemap use canonical urls for products, if the configuration option `catalog/seo/product_use_categories` is set to Yes.

This SEO setting, has an effect on `\Magento\Catalog\Model\Product\Url::getUrl` so it has a lot wider effect than just the sitemap.

Canonical URL tags on PDPs, however, don't take this setting into account and always produce urls without category paths. And rightfully so because otherwise we would be serving duplicate content.

The sitemap, I would argue, should direct search engines to the canonical url directly. Always. Not only based on this generic configuration setting. 

This PR aims to make the sitemap behave in a similar way to the canonical url generation.

### Fixed Issues (if relevant)

- Fixes #23363 New issues with product url's in sitemap.xml generation after PR 23129 got merged

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by AntonEvers** [comment]
@magento run all tests

File: app/code/Magento/Sitemap/Model/ResourceModel/Catalog/Product.php [hunk 1]
```diff
@@ -326,7 +322,7 @@ public function getCollection($storeId)
         )->joinLeft(
             ['url_rewrite' => $this->getTable('url_rewrite')],
             'e.entity_id = url_rewrite.entity_id AND url_rewrite.is_autogenerated = 1'
-            . $urlRewriteMetaDataCondition
+            . ' AND url_rewrite.target_path NOT LIKE "%/category/%"'
```
Comment (lbajsarowicz) [inline app/code/Magento/Sitemap/Model/ResourceModel/Catalog/Product.php]: That might kill the performance. 
Isn't there any other way to get rid of these URLs?

File: app/code/Magento/Sitemap/Model/ResourceModel/Catalog/Product.php [hunk 2]
```diff
@@ -84,11 +82,13 @@ class Product extends \Magento\Framework\Model\ResourceModel\Db\AbstractDb
 
     /**
      * @var \Magento\Catalog\Model\Product
+     * @deprecated 100.4.0 unused
```
Comment (lbajsarowicz) [inline app/code/Magento/Sitemap/Model/ResourceModel/Catalog/Product.php]: That are private properties, you can remove them without notice.

File: app/code/Magento/Sitemap/Model/ResourceModel/Catalog/Product.php [hunk 3]
```diff
@@ -84,11 +82,13 @@ class Product extends \Magento\Framework\Model\ResourceModel\Db\AbstractDb
...
      */
     private $productModel;
 
     /**
      * @var \Magento\Catalog\Helper\Image
+     * @deprecated 100.4.0 unused
```
Comment (lbajsarowicz) [inline app/code/Magento/Sitemap/Model/ResourceModel/Catalog/Product.php]: That are private properties, you can remove them without notice.

File: app/code/Magento/Sitemap/Model/ResourceModel/Catalog/Product.php [hunk 4]
```diff
@@ -101,6 +101,7 @@ class Product extends \Magento\Framework\Model\ResourceModel\Db\AbstractDb
      * Scope Config
      *
      * @var \Magento\Framework\App\Config\ScopeConfigInterface
+     * @deprecated 100.4.0 unused
```
Comment (lbajsarowicz) [inline app/code/Magento/Sitemap/Model/ResourceModel/Catalog/Product.php]: That are private properties, you can remove them without notice.

**Review by lbajsarowicz — CHANGES_REQUESTED** [review]
Please find another way to avoid `/category/` links as MySQL `LIKE` is going to kill the performance.

**Comment by hostep** [comment]
@AntonEvers: will this fix the "first problem" of https://github.com/magento/magento2/issues/23363 ?

**Comment by AntonEvers** [comment]
> @AntonEvers: will this fix the "first problem" of #23363 ?

@hostep before this PR:
```xml
<url>
<loc>https://magento2.test/category/product.html</loc>
<lastmod>2020-07-20T08:00:46+00:00</lastmod>
<changefreq>daily</changefreq>
<priority>1.0</priority>
</url>
```

After this PR:
```xml
<url>
<loc>https://magento2.test/product.html</loc>
<lastmod>2020-07-20T08:00:46+00:00</lastmod>
<changefreq>daily</changefreq>
<priority>1.0</priority>
</url>
```

So this indeed fixes  #23363

**Comment by AntonEvers** [comment]
@magento run all tests

**Comment by AntonEvers** [comment]
@magento run all tests

**Comment by AntonEvers** [comment]
@magento run all tests

**Comment by AntonEvers** [comment]
@magento run all tests

**Comment by AntonEvers** [comment]
@lbajsarowicz Functional Tests EE fails are not caused by this PR

**Review by sidolov — CHANGES_REQUESTED** [review]
@AntonEvers could you please cover changes with automated tests?

**Review by sidolov — DISMISSED** [review]

**Comment by magento-engcom-team** [comment]
Hi @sidolov, thank you for the review.
[ENGCOM-8071](https://jira.corp.magento.com/browse/ENGCOM-8071) has been created to process this Pull Request
:eight_spoked_asterisk: @sidolov, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Comment by sidolov** [comment]
The new behavior should be clarified with PO since it affects SEO.

**Comment by hostep** [comment]
@sidolov: please make sure the PO also reads all comments in https://github.com/magento/magento2/pull/23129 & https://github.com/magento/magento2/issues/23363 before forming an opinion.

Thanks!

**Comment by sivaschenko** [comment]
Approval Request: https://jira.corp.magento.com/browse/MC-38323

**Comment by JoostWan** [comment]
@AntonEvers Thanks for this fix. It's working fine but not for all products. In the case of our client there a some products with the wrong url in the sitemap with category paths. I checked the URL's in the database and the products with the wrong URL has 301 redirects in table `url_rewrite`. When i added some log to `public function getCollection($storeId)` the URL is good. But over here the URL is different with the categorie paths https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/Sitemap/Model/Sitemap.php#L462.

**Pull request merged**